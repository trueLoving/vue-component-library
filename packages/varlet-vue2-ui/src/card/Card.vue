<template>
  <div
    class="var-card"
    :class="[elevation ? `var-elevation--${elevation}` : 'var-elevation--2']"
    v-ripple="{ disabled: !ripple }"
    @click="handleClick"
  >
    <slot name="image">
      <img
        class="var-card__image"
        :style="{
          objectFit: fit,
          height: toSizeUnit(height),
        }"
        :src="src"
        :alt="alt"
        v-if="src"
      />
    </slot>
    <slot name="title">
      <div class="var-card__title" v-if="title">{{ title }}</div>
    </slot>
    <slot name="subtitle">
      <div class="var-card__subtitle" v-if="subtitle">{{ subtitle }}</div>
    </slot>
    <slot name="description">
      <div class="var-card__description" v-if="description">{{ description }}</div>
    </slot>
    <div class="var-card__footer" v-if="hasSlots('extra')">
      <slot name="extra" />
    </div>
  </div>
</template>

<script>
import Ripple from '../ripple'
import { defineComponent } from '../utils/create'
import { props } from './props'
import { toSizeUnit } from '../utils/elements'

export default defineComponent({
  name: 'VarCard',

  directives: { Ripple },

  props,

  data: () => ({
    toSizeUnit,
  }),

  methods: {
    handleClick(e) {
      const { disabled, getListeners } = this
      const { onClick } = getListeners()

      if (!onClick || disabled) {
        return
      }

      onClick(e)
    },
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import '../ripple/ripple';
@import './card';
</style>
