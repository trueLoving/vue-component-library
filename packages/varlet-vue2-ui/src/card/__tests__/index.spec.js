import example from '../example'
import Vue from 'vue'
import Card from '..'
import VarCard from '../Card'
import { mount } from '@vue/test-utils'

test('test card example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
})

test('test card plugin', () => {
  Vue.use(Card)
  expect(Vue.component(Card.name)).toBeTruthy()
})

test('test card props', async () => {
  const wrapper = mount(VarCard, {
    propsData: {
      title: 'This is Card',
      description: 'This is description',
      subtitle: 'This is subtitle',
      src: 'https://varlet-vue2.vercel.app/cat.jpg',
      fit: 'cover',
      height: '200px',
      alt: 'This is an image',
      elevation: '2',
      ripple: true,
    },
    scopedSlots: {
      extra: '<div>text</div>',
    },
  })

  expect(wrapper.find('img').attributes('style')).toMatch('height: 200px')
  expect(wrapper.find('img').attributes('style')).toMatch('object-fit: cover')
  expect(wrapper.find('img').attributes('alt')).toMatch('This is an image')
  expect(wrapper.find('img').attributes('src')).toMatch('https://varlet-vue2.vercel.app/cat.jpg')
  expect(wrapper.find('.var-card__title').text()).toBe('This is Card')
  expect(wrapper.find('.var-card__subtitle').text()).toBe('This is subtitle')
  expect(wrapper.find('.var-card__description').text()).toBe('This is description')
  expect(wrapper.find('.var-card__footer').text()).toBe('text')
  expect(wrapper.classes()).toContain('var-elevation--2')
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
})

test('test card onClick with null callback', () => {
  const wrapper = mount(VarCard)
  wrapper.trigger('click')
  wrapper.destroy()
})

test('test card onClick', () => {
  const onClick = jest.fn()
  const wrapper = mount(VarCard, {
    listeners: {
      click: onClick,
    },
  })

  wrapper.trigger('click')
  expect(onClick).toHaveBeenCalledTimes(1)
  wrapper.destroy()
})
