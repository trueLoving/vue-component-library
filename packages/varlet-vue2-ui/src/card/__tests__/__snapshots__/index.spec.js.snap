// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test card example 1`] = `
"<div class=\\"var-card-example\\">
  <div class=\\"app-type\\">基本使用</div>
  <div class=\\"var-card var-elevation--2\\">
    <!---->
    <div class=\\"var-card__title\\">风景</div>
    <!---->
    <div class=\\"var-card__description\\">公园的树林也很美。在公园的小山上栽满了树木，梧桐树的叶子随着时间的流逝慢慢变黄，纷纷飘落；枫树的叶子却变红了，公园笼罩在片片红云中，也使秋天增添了一分热情。而柏树的叶子仍是那么青翠欲滴，令你陶醉极了。山上有一群孩子在快乐的嬉戏，不时传来阵阵欢笑声，瞧，他们玩得多起劲呀，给树林增添了活力。</div>
    <!---->
  </div>
  <div class=\\"app-type\\">显示副标题</div>
  <div class=\\"var-card var-elevation--2\\">
    <!---->
    <div class=\\"var-card__title\\">风景</div>
    <div class=\\"var-card__subtitle\\">公园里的风景</div>
    <div class=\\"var-card__description\\">公园的树林也很美。在公园的小山上栽满了树木，梧桐树的叶子随着时间的流逝慢慢变黄，纷纷飘落；枫树的叶子却变红了，公园笼罩在片片红云中，也使秋天增添了一分热情。而柏树的叶子仍是那么青翠欲滴，令你陶醉极了。山上有一群孩子在快乐的嬉戏，不时传来阵阵欢笑声，瞧，他们玩得多起劲呀，给树林增添了活力。</div>
    <!---->
  </div>
  <div class=\\"app-type\\">显示图片</div>
  <div class=\\"var-card var-elevation--2\\"><img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" class=\\"var-card__image\\" style=\\"object-fit: cover;\\">
    <div class=\\"var-card__title\\">风景</div>
    <div class=\\"var-card__subtitle\\">公园里的风景</div>
    <div class=\\"var-card__description\\">公园的树林也很美。在公园的小山上栽满了树木，梧桐树的叶子随着时间的流逝慢慢变黄，纷纷飘落；枫树的叶子却变红了，公园笼罩在片片红云中，也使秋天增添了一分热情。而柏树的叶子仍是那么青翠欲滴，令你陶醉极了。山上有一群孩子在快乐的嬉戏，不时传来阵阵欢笑声，瞧，他们玩得多起劲呀，给树林增添了活力。</div>
    <!---->
  </div>
  <div class=\\"app-type\\">使用插槽</div>
  <div class=\\"var-card var-elevation--2\\"><img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" class=\\"var-card__image\\" style=\\"object-fit: cover;\\">
    <div class=\\"var-card__title\\">风景</div>
    <div class=\\"var-card__subtitle\\">公园里的风景</div>
    <div class=\\"var-card__description\\">公园的树林也很美。在公园的小山上栽满了树木，梧桐树的叶子随着时间的流逝慢慢变黄，纷纷飘落；枫树的叶子却变红了，公园笼罩在片片红云中，也使秋天增添了一分热情。而柏树的叶子仍是那么青翠欲滴，令你陶醉极了。山上有一群孩子在快乐的嬉戏，不时传来阵阵欢笑声，瞧，他们玩得多起劲呀，给树林增添了活力。</div>
    <div class=\\"var-card__footer\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--primary var-elevation--2\\" style=\\"margin-right: 10px;\\">
        <!---->
        <div class=\\"var-button__content\\">添加按钮</div>
      </button> <button class=\\"var-button var--box var-button--normal var--inline-flex var-button--warning var-elevation--2\\">
        <!---->
        <div class=\\"var-button__content\\">添加按钮</div>
      </button></div>
  </div>
  <div class=\\"app-type\\">水波效果</div>
  <div class=\\"var-card var-elevation--2\\">
    <!---->
    <div class=\\"var-card__title\\">风景</div>
    <div class=\\"var-card__subtitle\\">公园里的风景</div>
    <div class=\\"var-card__description\\">公园的树林也很美。在公园的小山上栽满了树木，梧桐树的叶子随着时间的流逝慢慢变黄，纷纷飘落；枫树的叶子却变红了，公园笼罩在片片红云中，也使秋天增添了一分热情。而柏树的叶子仍是那么青翠欲滴，令你陶醉极了。山上有一群孩子在快乐的嬉戏，不时传来阵阵欢笑声，瞧，他们玩得多起劲呀，给树林增添了活力。</div>
    <!---->
  </div>
</div>"
`;

exports[`test card props 1`] = `
"<div class=\\"var-card var-elevation--2\\"><img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" alt=\\"This is an image\\" class=\\"var-card__image\\" style=\\"object-fit: cover; height: 200px;\\">
  <div class=\\"var-card__title\\">This is Card</div>
  <div class=\\"var-card__subtitle\\">This is subtitle</div>
  <div class=\\"var-card__description\\">This is description</div>
  <div class=\\"var-card__footer\\">
    <div>text</div>
  </div>
</div>"
`;
